azure-iot-device/azure/iot/device/provisioning/pipeline/mqtt_topic.py

```diff
@@ -0,0 +1,92 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import logging
+import six.moves.urllib as urllib
+
+logger = logging.getLogger(__name__)
+
+
+def _get_topic_base():
+    """
+    return the string that creates the beginning of all topics for DPS
+    """
+    return "$dps/registrations/"
+
+
+def get_topic_for_subscribe():
+    """
+    return the topic string used to subscribe for receiving future responses from DPS
+    """
+    return _get_topic_base() + "res/#"
+
+
+def get_topic_for_register(rid):
+    """
+    return the topic string used to publish telemetry
+    """
+    return (_get_topic_base() + "PUT/iotdps-register/?$rid={rid}").format(rid=rid)
+
+
+def get_topic_for_query(rid, operation_id):
+    """
+    :return: The topic for cloud to device messages.It is of the format
+    "devices/<deviceid>/messages/devicebound/#"
+    """
+    return (
+        _get_topic_base() + "GET/iotdps-get-operationstatus/?$rid={rid}&operationId={operation_id}"
+    ).format(rid=rid, operation_id=operation_id)
+
+
+def get_topic_for_response():
+    """
+    return the topic string used to publish telemetry
+    """
+    return _get_topic_base() + "res/"
+
+
+def is_query_topic(topic):
+    if "GET/iotdps-get-operationstatus" in topic:
+        return True
+    return False
+
+
+def is_dps_response_topic(topic):
+    """
+    Topics for responses from DPS are of the following format:
+    $dps/registrations/res/<statuscode>/?$<key1>=<value1>&<key2>=<value2>&<key3>=<value3>
+    :param topic: The topic string
+    """
+    if get_topic_for_response() in topic:
+        return True
+    return False
+
+
+def extract_properties_from_topic(topic):
```

Comment:
Not sure how relevant this is here, but in the iothub client there's an issue where this function has a generic name, but is specifically tied to a certain type of topic string (messages), and needs to be refactored as a result.

If there are different types of possible topic strings for DPS, this may present some kind of problem for you? Or not. I'm not quite sure the scenarios here. If this is working fine, leave it alone I guess.

When we refactor how this stuff all happens in the IoTHubClient, depending on the approach, we may end up bringing some of the changes over here to reduce duplicate code.

Reply:
as far as I know this would cover all topic strings for DPS

---
In reply to: [287548205](https://github.com/Azure/azure-iot-sdk-python-preview/pull/97#discussion_r287548205) [](ancestors = 287548205)